quantization/LowPrecisionQuantization.h

```diff
@@ -0,0 +1,225 @@
+      return;
+    }
+    _mean_vector.resize(_dimension);
+
+    omp_set_num_threads(1);
+
+#pragma omp parallel for default(none) shared(vectors, num_vectors)
+    for (uint64_t vec_index = 0; vec_index < num_vectors; vec_index++) {
+      for (uint32_t dim_index = 0; dim_index < _dimension; dim_index++) {
+#pragma omp atomic
+        _mean_vector[dim_index] += vectors[vec_index * _dimension + dim_index];
+      }
+    }
+    std::cout << "[info] LPQ: computing the mean vector" << std::endl;
+    for (uint32_t dim_index = 0; dim_index < _dimension; dim_index++) {
+      _mean_vector[dim_index] /= num_vectors;
+    }
+
+    _is_trained = true;
+  }
+
+private:
+  size_t _num_bits;
+  size_t _dimension;
+  size_t _data_size_bytes;
+
+  bool _is_trained;
+  METRIC_TYPE _metric_type;
+
+  // This is a vector along the dimensions of the entire vectors.
+  // mu_i is the mean of the ith dimension.
+  std::vector<float> _mean_vector;
+
+  std::variant<SquaredL2Distance, InnerProductDistance> _distance;
+
+  friend class cereal::access;
+
+  template <typename Archive> void serialize(Archive &ar) {
+    ar(_num_bits, _dimension, _data_size_bytes, _is_trained, _metric_type,
+       _mean_vector);
+
+    if (Archive::is_loading::value) {
+      if (_metric_type == METRIC_TYPE::EUCLIDEAN) {
+        _distance = SquaredL2Distance(_dimension);
+      } else if (_metric_type == METRIC_TYPE::INNER_PRODUCT) {
+        _distance = InnerProductDistance(_dimension);
+      }
+    }
+  }
+
+  /**
+   * @brief Computes the upper and lower limits of the quantization range.
+   * for a given vector.
+   *
+   * @param vector data vector of size _dimension
+   */
+  std::pair<float, float> getMinMax(const float *vector) const {
+    assert(_is_trained);
+
+    float min = std::numeric_limits<float>::max();
+    float max = std::numeric_limits<float>::min();
+
+    for (uint32_t dim_index = 0; dim_index < _dimension; dim_index++) {
+      auto diff = vector[dim_index] - _mean_vector[dim_index];
+      if (diff < min) {
+        min = diff;
+      }
+      if (diff > max) {
+        max = diff;
+      }
+    }
+    return {min, max};
+  }
+
+  /**
+   * @brief Quantizes a given vector following the locally-adaptive quantization
+   * scheme. quantized_vector must be pre-allocated.
+   *
+   * @param vector
+   * @param quantized_vector
+   */
+  template <typename T>
+  void quantizeVector(const float *vector, T *quantized_vector) const {
+    assert(_is_trained);
+    assert(sizeof(T) == _num_bits / 8);
+
+    auto [min, max] = getMinMax(vector);
+
+    float delta = (max - min) / ((1 << _num_bits) - 1);
+
+    for (uint32_t dim_index = 0; dim_index < _dimension; dim_index++) {
+      auto first_term = (vector[dim_index] - min) / delta;
+      first_term += 0.5;
+      quantized_vector[dim_index] =
+          static_cast<T>(delta * std::floor(first_term) + min);
+    }
+  }
+
+  template <typename data_t>
+  float distanceImpl(const void *x, const void *y,
```

Comment:
This can be cleaned up!

Reply:
I would suggest having the asymmetric be specified at the class-level (either as a template, init arg, or assumption of the class, e.g. "LowPrecisionQuantizerAsym" and not an arg to distanceImpl), since it's unlikely that you will call both symmetric and asymmetric options on the same DistanceInterface with conditions that are determined at runtime.